### PR TITLE
Supports toLower() and toUpper() for non-english characters

### LINKED ISSUE #2670: toLower() and toUpper() do not change the case of non-English letters
Greek:

```
GRAPH.QUERY g "RETURN toLower('ΑαΒβΓγΔδΕεΖζΗηΘθΙιΚκΛλΜμΝνΞξΟοΠπΡρΣσςΤτΥυΦφΧχΨψΩω')"
```
Expected: "ααββγγδδεεζζηηθθιικκλλμμννξξοοππρρσσςττυυφφχχψψωω"

```
GRAPH.QUERY g "RETURN toUpper('ΑαΒβΓγΔδΕεΖζΗηΘθΙιΚκΛλΜμΝνΞξΟοΠπΡρΣσςΤτΥυΦφΧχΨψΩω')"
```
Expected: "ΑΑΒΒΓΓΔΔΕΕΖΖΗΗΘΘΙΙΚΚΛΛΜΜΝΝΞΞΟΟΠΠΡΡΣΣΣΤΤΥΥΦΦΧΧΨΨΩΩ"

Cyrillic:

```
GRAPH.QUERY g "RETURN toLower('АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНнОоПпРрСсТтУуФфХхЦцЧчШшЩщЬьЭэЮюЯя')
```
Expected: "ааббввггддеежжззииййккллммннооппррссттууффххццччшшщщььээююяя"

```
GRAPH.QUERY g "RETURN toUpper('АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНнОоПпРрСсТтУуФфХхЦцЧчШшЩщЬьЭэЮюЯя')
```
Expected: "ААББВВГГДДЕЕЖЖЗЗИИЙЙККЛЛММННООППРРССТТУУФФХХЦЦЧЧШШЩЩЬЬЭЭЮЮЯЯ"

but currently, no conversion is done.

### PR DESCRIPTION
**Pull request by stevealexr**
Fixes #2670 

- added optional locale parameter
- used some flags to disable unused icu features but the documentation is not very clear
- added some tests but they don't run so I tested them manually. Tested on ubuntu 22.04 and partially tested on mac monterey 12.6.1.
- The icu library toUpper and toLower functions may return error but I am not sure what is the preferable approach for error handling here so I ignore those errors for now.

### CODE REVIEW

File: tests/flow/test_function_calls.py [hunk 1]
```diff
@@ -2223,3 +2223,21 @@ def test87_typeof(self):
...
+            "RETURN toLower('TITLE', 'tr')": [['tıtle']], # Locale-sensitive
+            "RETURN toLower('TITLE', 'en')": [['title']]
+        }
+        for query, expected_result in query_to_expected_result.items():
+            self.get_res_and_assertEquals(query, expected_result)
+    
+    def test89_ToUpperUnicode(self):
+        query_to_expected_result = {
+            "RETURN toUpper('αβγδεζηθικλμνξοπρστυφχψω')": [['ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩ']], # greek
+            "RETURN toUpper('абвгдеёжзийклмнопрстуфхцчшщъыьэюя')": [['АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ']], #cyrillic
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:2240]: For consistency please use the same strings as in `test88_ToLowerUnicode`

File: src/Makefile [hunk 2]
```diff
@@ -308,6 +317,25 @@ endif
...
+$(ICU4C_LIBS):
+ifeq (,$(wildcard $(ICU4C_LIBS)))
+	@mkdir -p $(ICU4C_MAKE_DIR) ;\
+	cd $(ICU4C_MAKE_DIR) ;\
+	chmod +x $(ICU4C_SRC_DIR)/runConfigureICU $(ICU4C_SRC_DIR)configure $(ICU4C_SRC_DIR)install-sh ;\
+	CPPFLAGS="-fPIC -DU_NO_DEFAULT_INCLUDE_UTF_HEADERS=1 -DUCONFIG_ONLY_COLLATION=1" \
+	$(ICU4C_SRC_DIR)/runConfigureICU $(ICU_CONFIG_OS) --enable-static --disable-shared \
+	prefix=$(ICU4C) ;\
+	$(MAKE);\
+	$(MAKE) install
```
Comment (swilly22) [inline src/Makefile:336]: Why do we to `make install` isn't `make` enough ?

File: src/Makefile [hunk 3]
```diff
@@ -318,7 +346,7 @@ endef
 
 # Build the module...
 $(TARGET): $(MODULE_DEPS)
-	$(REDISGRAPH_CC) -o $@ $(CC_OBJECTS) $(LIBS) $(SHOBJ_LDFLAGS) -lc -lm
+	$(REDISGRAPH_CC) -o $@ $(CC_OBJECTS) $(LIBS) $(SHOBJ_LDFLAGS) -lc -lm -lstdc++
```
Comment (swilly22) [inline src/Makefile:349]: Why was `-lstdc++` added?

File: src/Makefile [hunk 4]
```diff
@@ -336,6 +364,11 @@ ifeq ($(ALL),1)
 	@$(MAKE) -C $(ROOT)/deps/libcypher-parser clean
 	@cd $(ROOT)/deps/libcypher-parser; git ls-files -o | xargs rm
 	@$(MAKE) -C $(REDISEARCH_DIR) clean ALL=1 BINROOT=$(REDISEARCH_BINROOT)
+
+	@$(MAKE) -C $(ICU4C_MAKE_DIR) clean
+	@$(MAKE) -C $(ICU4C_MAKE_DIR) distclean
```
Comment (swilly22) [inline src/Makefile:369]: isn't `distclean` enough ?

File: .gitmodules [hunk 5]
```diff
@@ -16,3 +16,6 @@
 [submodule "deps/readies"]
 	path = deps/readies
 	url = https://github.com/RedisLabsModules/readies.git
+[submodule "deps/icu"]
+	path = deps/icu
+	url = https://github.com/unicode-org/icu.git
```
Comment (swilly22) [inline .gitmodules:21]: @LiorKogan @gkorland please review project's license

File: src/util/strutil.c [hunk 6]
```diff
@@ -8,31 +8,57 @@
...
 #include "RG.h"
 #include "rmalloc.h"
+#include "../../build/icu/build/include/unicode/ustring.h"
 
-void str_tolower(const char *str, char *lower, size_t *lower_len) {
+static void str_changecase(const char *str, char *result, size_t *result_len, const char *locale, bool toUpper) {
 	size_t str_len = strlen(str);
 	//Avoid overflow
-	ASSERT(*lower_len >= str_len);
+	assert(*result_len >= str_len);
```
Comment (swilly22) [inline src/util/strutil.c:16]: Please use the upper case version of assert `ASSERT`

File: src/util/strutil.c [hunk 7]
```diff
@@ -8,31 +8,57 @@
 #include <ctype.h>
 #include "RG.h"
 #include "rmalloc.h"
+#include "../../build/icu/build/include/unicode/ustring.h"
 
-void str_tolower(const char *str, char *lower, size_t *lower_len) {
+static void str_changecase(const char *str, char *result, size_t *result_len, const char *locale, bool toUpper) {
```
Comment (swilly22) [inline src/util/strutil.c:13]: Please add a descriptive comment for this function
```c
// description...
static void str_changecase
(
    const char *str,  //...
    char *result, size_t *result_len, //...
    const char *locale, //...
    bool toUpper  //...
) {
```

File: src/util/strutil.c [hunk 8]
```diff
@@ -8,31 +8,57 @@
...
-	//Avoid overflow
-	ASSERT(*upper_len >= str_len);
+    u_uastrcpy(original, str);
+
+    size_t size;
+	if (toUpper) {
+		size = u_strToUpper(converted, u_result_len, original, u_og_len, locale, &errorCode);
+	} else {
+		size = u_strToLower(converted, u_result_len, original, u_og_len, locale, &errorCode);
+	}
```
Comment (swilly22) [inline src/util/strutil.c:34]: indentation

File: src/util/strutil.c [hunk 9]
```diff
@@ -8,31 +8,57 @@
...
 
-	//Update lower len
-	*upper_len = str_len;
+    // if the result is larger than allocated length
+    if (size > u_result_len) {
```
Comment (swilly22) [inline src/util/strutil.c:37]: Can you give an example for when `size > u_result_len` is true?

File: src/util/strutil.c [hunk 10]
```diff
@@ -8,31 +8,57 @@
...
+		*result_len = size;
+        if (toUpper) {
+			u_strToUpper(converted, u_result_len, original, u_og_len, locale, &errorCode);
+		} else {
+			u_strToLower(converted, u_result_len, original, u_og_len, locale, &errorCode);
+		}
+    }
+	if (U_FAILURE(errorCode)) {
+		// do something to handle error
+		*result = *str;
```
Comment (swilly22) [inline src/util/strutil.c:48]: in case of a raise a run-time exception and set `result` to NULL

File: src/util/strutil.c [hunk 11]
```diff
@@ -8,31 +8,57 @@
...
+	} else {
+		u_austrcpy(result, converted);
```
Comment (swilly22) [inline src/util/strutil.c:50]: Please add a short comment explaining what this function does
e.g.
`u_austrcpy(result, converted); // copy unicode to ascii `

File: src/util/strutil.c [hunk 12]
```diff
@@ -8,31 +8,57 @@
...
-}
+	UChar* original = (UChar *) rm_malloc((u_og_len + 1) * sizeof(UChar));
+    UChar* converted = (UChar *) rm_malloc((u_result_len + 1) * sizeof(UChar));
+    UErrorCode errorCode = U_ZERO_ERROR;
 
-void str_toupper(const char *str, char *upper, size_t *upper_len) {
-	size_t str_len = strlen(str);
-	//Avoid overflow
-	ASSERT(*upper_len >= str_len);
+    u_uastrcpy(original, str);
```
Comment (swilly22) [inline src/util/strutil.c:27]: Please add a short comment explaining what this function does
e.g.
`u_uastrcpy(original, str); // copy ascii to unicode`

File: src/util/strutil.c [hunk 13]
```diff
@@ -8,31 +8,57 @@
...
-	*lower_len = str_len;
+	size_t u_og_len = str_len;
+	size_t u_result_len = str_len;
 
-	size_t i = 0;
-	for(; i < str_len; i++) lower[i] = tolower(str[i]);
-	lower[i] = 0;
-}
+	UChar* original = (UChar *) rm_malloc((u_og_len + 1) * sizeof(UChar));
+    UChar* converted = (UChar *) rm_malloc((u_result_len + 1) * sizeof(UChar));
```
Comment (swilly22) [inline src/util/strutil.c:24]: Can't we use a single variable / allocation?
```c
UChar* converted = (UChar *) rm_malloc((u_result_len + 1) * sizeof(UChar));
u_uastrcpy(converted, str);
size = u_strToUpper(converted, u_result_len, converted, u_og_len, locale, &errorCode);
```

File: src/util/strutil.h [hunk 14]
```diff
@@ -6,11 +6,11 @@
 
 #pragma once
 
-//convert str to a lower case string and save it in lower
-void str_tolower(const char *str, char *lower, size_t *lower_len);
+//convert str to a lower case string and save it in lower, pass "" for the root locale or NULL for the default locale
+void str_tolower(const char *str, char *lower, size_t *lower_len, const char *locale);
```
Comment (swilly22) [inline src/util/strutil.h:10]: Due to the overhead and the distinction between load-time and runtime invocations
I prefer if we introduce two additional functions:
```
void str_tolower(const char *str, char *lower, size_t *lower_len);
void str_tolower_locale(const char *str, char *lower, size_t *lower_len, const char *locale);
void str_toupper(const char *str, char *upper, size_t *upper_len);
void str_toupper_locale(const char *str, char *upper, size_t *upper_len, const char *locale);
```

The original `str_toupper` and `str_tolower` should be called when we explicitly pass NULL for locale in this PR.
e.g.
`str_tolower(procedure, lowercase_proc_name, &lowercase_proc_name_len, NULL);` -> `str_tolower(procedure, lowercase_proc_name, &lowercase_proc_name_len);`

Calls to `toUpper` or `toLower` made by the user at runtime will use the new `str_tolower_locale` and `str_toupper_locale` versions.

**Review by swilly22 — CHANGES_REQUESTED** [review]
Thank you, please see comments.

File: .gitmodules (hunk 5 above)
Comment (LiorKogan) [reply .gitmodules:21]: We are good with the license.  The only requirement is that either
(a) this copyright and permission notice appear with all copies of the Data Files or Software, or
(b) this copyright and permission notice appear in associated Documentation.

File: src/Makefile (hunk 2 above)
Comment (stevealexr) [reply src/Makefile:336]: I am following the instruction [here](https://unicode-org.github.io/icu/userguide/icu4c/build.html#how-to-build-and-install-on-unix). I think `make` compiles the files while `make install` copies the compiled files to the predefined directory.

File: src/Makefile (hunk 3 above)
Comment (stevealexr) [reply src/Makefile:349]: The icu4c library requires c++ library even for its c api.

File: src/Makefile (hunk 4 above)
Comment (stevealexr) [reply src/Makefile:369]: Yes, `distclean` alone is enough.

File: src/Makefile (hunk 3 above)
Comment (LiorKogan) [reply src/Makefile:349]: @stevealexr we don't want to add C++ dependencies to RedisGraph. This is why we requested "minimal subset of the latest ICU library, converted from C++ to C".

File: src/util/strutil.c (hunk 8 above)
Comment (stevealexr) [reply src/util/strutil.c:34]: I think some files use tab as indentation while some use spaces. I changed all spaces to tab here to be consistent.

File: src/util/strutil.c (hunk 9 above)
Comment (stevealexr) [reply src/util/strutil.c:37]: The `u_result_len` is equal to the original string length. German `ß` to upper case is `SS` and it is larger.

Comment (stevealexr) [reply src/util/strutil.c:37]: Some characters are encoded differently when they change case. German `ß` to upper case becomes `SS` and it is larger.

I tested it with this code 
```c
UErrorCode errorCode = U_ZERO_ERROR;
UChar o[100];
UChar new[100];

char ochar[100];
char newchar[100];
u_uastrcpy(o, "ß");
u_austrcpy(ochar, o);
u_strToUpper(new, 100, o, 1, "", &errorCode);
u_austrcpy(newchar, new);
printf("uchar length: %s, %d \n", ochar, u_strlen(o));
printf("uchar length: %s, %d \n", newchar, u_strlen(new));
```
and the result is 
```
uchar length: ß, 1 
uchar length: SS, 2
```

File: src/Makefile (hunk 3 above)
Comment (stevealexr) [reply src/Makefile:349]: I will try to convert relevant functions and their dependencies to C.

**Comment by stevealexr** [comment]
I am quite busy so I will stop working on this. Feel free to pick up from where I left off, the majority of the remaining work is porting icu4c to c. You also need to switch icu4c to the c port and implement those suggestions above.

**Pull request closed without merging**